R/likelihood.R

```diff
@@ -0,0 +1,170 @@
+#' individual == FALSE and \code{0 <= obs_prob < 1} (imperfect observation).
+#'
+#' If \code{log = FALSE}, the same structure of outputs as above are returned,
+#' except that likelihoods, instead of log-likelihoods, are calculated in all
+#' cases. Moreover, the joint likelihoods are the product, instead of the sum,
+#' of the individual likelihoods.
+#' @seealso offspring_ll(), pois_size_ll(), nbinom_size_ll(), gborel_size_ll(),
+#' pois_length_ll(), geom_length_ll()
+#' @author Sebastian Funk
+#' @examples
+#' # example of observed chain sizes
+#' set.seed(121)
+#' # randomly generate 20 chains of size 1 to 10
+#' chain_sizes <- sample(1:10, 20, replace = TRUE)
+#' likelihood(
+#'   chains = chain_sizes, statistic = "size",
+#'   offspring_dist = "pois", nsim_obs = 100, lambda = 0.5
+#' )
+#' @export
+likelihood <- function(chains, statistic = c("size", "length"), offspring_dist,
+                       nsim_obs, log = TRUE, obs_prob = 1, stat_max = Inf,
+                       exclude = NULL, individual = FALSE, ...) {
+  statistic <- match.arg(statistic)
+
+  ## Input checking
+  ## Check nsim_obs when specified
+  if (!missing(nsim_obs)) {
+    checkmate::assert_number(
+      nsim_obs, lower = 1, finite = TRUE, na.ok = FALSE
+    )
+  }
+
+  checkmate::assert_numeric(
+    chains, lower = 0, upper = Inf, any.missing = FALSE
+  )
+  checkmate::assert_character(statistic)
+  check_offspring_valid(offspring_dist)
+  checkmate::assert_number(
+    obs_prob, lower = 0, upper = 1, finite = TRUE, na.ok = FALSE
+  )
+  checkmate::assert_number(
+    stat_max, lower = 0, na.ok = FALSE
+  )
+  checkmate::assert_logical(
+    log, any.missing = FALSE, all.missing = FALSE, len = 1
+  )
+  checkmate::assert_logical(
+    individual, any.missing = FALSE, all.missing = FALSE, len = 1
+  )
+  checkmate::assert_numeric(
+    exclude, null.ok = TRUE
+  )
+
+  if (obs_prob < 1) {
+    if (missing(nsim_obs)) {
+      stop("'nsim_obs' must be specified if 'obs_prob' is < 1")
+    }
+
+    statistic_func <- get_statistic_func(statistic)
+
+    stat_rep_list <- replicate(nsim_obs, pmin(
+      statistic_func(
+        length(chains),
+        chains, obs_prob
+      ),
+      stat_max
+    ), simplify = FALSE)
+    stat_rep_vect <- unlist(stat_rep_list)
+    if (!is.finite(stat_max)) {
+      stat_max <- max(stat_rep_vect) + 1
+    }
+  } else {
+    chains[chains >= stat_max] <- stat_max
+    stat_rep_vect <- chains
+    stat_rep_list <- list(chains)
+  }
+
+  ## determine for which sizes to calculate the log-likelihood
+  ## (for true chain size)
+  if (any(stat_rep_vect == stat_max)) {
+    calc_sizes <- seq_len(stat_max - 1)
+  } else {
+    calc_sizes <- unique(c(stat_rep_vect, exclude))
+  }
+
+  ## get log-likelihood function as given by offspring_dist and statistic
+  likelihoods <- vector(mode = "numeric")
+  ll_func <- construct_offspring_ll_name(offspring_dist, statistic)
+  pars <- as.list(unlist(list(...))) ## converts vectors to lists
```

Comment:
This a case of a legacy code that I didn't refactor because I wasn't sure what it would break. I always thought it could be written as `as.list(c(...))` or with a base R equivalent of `rlang::list2()`. What do you suggest?